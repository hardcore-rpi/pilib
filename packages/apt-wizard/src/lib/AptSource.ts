export enum DebTypeEnum {
  deb = 'deb',
  debSrc = 'deb-src',
}

export enum DebVersionEnum {
  buster = 'buster',
  stretch = 'stretch',
  jessie = 'jessie',
  wheezy = 'wheezy',
}

export enum DebRepoEnum {
  main = 'main',
  contrib = 'contrib',
  nonFree = 'non-free',
  firmware = 'firmware',
  ui = 'ui',
}

export abstract class AptBaseSource {
  abstract readonly name: string;
  abstract readonly title: string;
  abstract readonly note: string;
  abstract readonly sources: {
    debType: DebTypeEnum;
    url: string;
    version: DebVersionEnum;
    repoList: DebRepoEnum[];
  }[];

  toFullSourceUrls(): string[] {
    return this.sources.map(s => [s.debType, s.url, s.version, ...s.repoList].join(' '));
  }
}

/** 清华大学开源镜像 */
export class AptTunaSource extends AptBaseSource {
  name = 'tuna';
  title = '清华大学开源软件镜像站';
  note = 'https://mirrors.tuna.tsinghua.edu.cn/help/raspbian/';
  sources = [
    {
      debType: DebTypeEnum.deb,
      url: 'http://mirrors.tuna.tsinghua.edu.cn/raspbian/raspbian/',
      version: DebVersionEnum.buster,
      repoList: [DebRepoEnum.main, DebRepoEnum.nonFree, DebRepoEnum.contrib],
    },
    {
      debType: DebTypeEnum.debSrc,
      url: 'http://mirrors.tuna.tsinghua.edu.cn/raspbian/raspbian/',
      version: DebVersionEnum.buster,
      repoList: [DebRepoEnum.main, DebRepoEnum.nonFree, DebRepoEnum.contrib],
    },
    {
      debType: DebTypeEnum.deb,
      url: 'http://mirrors.tuna.tsinghua.edu.cn/raspberrypi/',
      version: DebVersionEnum.buster,
      repoList: [DebRepoEnum.main, DebRepoEnum.ui],
    },
  ];
}

export class AptSourceCollection {
  list: AptBaseSource[] = [new AptTunaSource()];

  getByName(name: string) {
    return this.list.find(s => s.name === name);
  }
}
