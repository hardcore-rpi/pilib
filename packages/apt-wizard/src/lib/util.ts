import { IFS } from '../type/fs';

export function mkdirp(path: string, fs: IFS) {
  const lastSlashIndex = path.lastIndexOf('/');

  // 这里的 || 是为了保留根路径
  const parentPath = path.slice(0, lastSlashIndex) || '/';

  // 如果父目录不在，就先创建父目录
  if (!fs.existsSync(parentPath)) mkdirp(parentPath, fs);
  if (!fs.existsSync(path)) fs.mkdirSync(path);
}

export function getSystemInfo(fs: IFS) {
  let releaseCodeName: string = '';

  if (fs.existsSync('/etc/os-release')) {
    const content: string = fs.readFileSync('/etc/os-release', { encoding: 'utf-8' });
    releaseCodeName = (content.match(/^VERSION_CODENAME=(.*)\n/) || [])[1];
  }

  return { releaseCodeName };
}
