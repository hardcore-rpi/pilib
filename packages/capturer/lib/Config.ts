export class Config {
  readonly CAMERA_ID: number = +(process.env.CAMERA_ID || '0');
  readonly CAMERA_WIDTH: number = +(process.env.CAMERA_WIDTH || '320');
  readonly CAMERA_HEIGHT: number = +(process.env.CAMERA_HEIGHT || '240');
  readonly UPLOAD_ENDPOINT: string = process.env.UPLOAD_ENDPOINT || './output/';

  toLogStr() {
    return Object.entries(this)
      .map(([n, v]) => `${n}=${v}`)
      .join(' ');
  }
}
