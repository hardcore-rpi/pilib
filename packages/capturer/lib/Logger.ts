export class Logger {
  constructor(readonly id: string) {}

  private ft(msg: string) {
    const timestamp = new Date().toLocaleString('zh-cn');

    return this.id ? `[${timestamp}][${this.id}] ${msg}` : `[${timestamp}] ${msg}`;
  }

  extend(id: string) {
    return new Logger([this.id, id].join('.'));
  }

  info(msg: string) {
    console.info(this.ft(msg));
  }

  error(msg: string) {
    console.error(this.ft(msg));
  }
}
