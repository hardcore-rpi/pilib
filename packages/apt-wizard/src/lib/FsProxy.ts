import { IFS } from '../type/fs';
import * as pathUtil from 'path';
import { mkdirp } from './util';

export enum WriteCacheEnum {
  write = 'write',
}

export class FsProxy implements IFS {
  private writeCache: {
    [path: string]: {
      type: WriteCacheEnum;
      content?: Buffer;
    };
  } = {};

  constructor(private readonly diskFs: IFS) {}

  readFileSync: IFS['readFileSync'] = (...args) => {
    const [path, opt] = args;

    // 先从缓存中找
    if (this.writeCache[path]) {
      return opt?.encoding
        ? this.writeCache[path].content!.toString(opt.encoding)
        : this.writeCache[path];
    }

    return this.diskFs.readFileSync(...args);
  };

  writeFileSync: IFS['writeFileSync'] = (...args) => {
    // 同步写入缓存
    const [path, data, opt] = args;
    this.writeCache[path] = {
      type: WriteCacheEnum.write,
      content: Buffer.from(data, opt?.encoding),
    };

    return this.writeCache[path].content;
  };

  existsSync: IFS['existsSync'] = (...args) => {
    const [path] = args;

    // 先从缓存中找
    return !!this.writeCache[path] || this.diskFs.existsSync(...args);
  };

  mkdirSync: IFS['mkdirSync'] = () => {};

  getWriteCache() {
    return { ...this.writeCache };
  }

  /** 写 cache 到 fs */
  flushWriteCache(
    backupDir: string | null,
    writePathPrefix: string = '',
    targetFs: IFS = this.diskFs
  ) {
    // 从 diskFs 备份
    if (backupDir !== null) {
      const cachePaths = Object.keys(this.writeCache);

      cachePaths.forEach(path => {
        // 跳过不存的路径（比如 cache 中新增的）
        if (!this.diskFs.existsSync(path)) return;

        const originContent = this.diskFs.readFileSync(path);
        const writePath = pathUtil.join(backupDir, path);

        mkdirp(pathUtil.dirname(writePath), targetFs);

        targetFs.writeFileSync(writePath, originContent);
      });
    }

    // 写 cache 到 targetFs
    Object.entries(this.writeCache).forEach(([path, spec]) => {
      const writePath = pathUtil.join(writePathPrefix, path);

      mkdirp(pathUtil.dirname(writePath), targetFs);

      switch (spec.type) {
        case WriteCacheEnum.write:
          targetFs.writeFileSync(writePath, spec.content);
          break;
        default:
          break;
      }
    });
  }
}
