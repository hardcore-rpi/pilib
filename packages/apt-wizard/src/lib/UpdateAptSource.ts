import { IFS } from '../type/fs';
import * as pathUtil from 'path';
import { AptBaseSource } from './AptSource';
import { mkdirp } from './util';

/** 更新软件源 */
export class UpdateAptSource {
  constructor(
    private readonly config: {
      /** 写入路径 */
      aptSourceWritePath: string;

      /** 删除文件 */
      cleanPaths: string[];
    },
    private readonly fs: IFS
  ) {}

  /** 清除和备份 */
  private cleanAndBackUp() {
    const cleanFiles: {
      path: string;
      content: string;
    }[] = [];

    // 收集备份文件(要把 aptSourceWritePath 也加入收集列表)
    const cleanPathSet = new Set([...this.config.cleanPaths, this.config.aptSourceWritePath]);

    cleanPathSet.forEach(path => {
      if (!this.fs.existsSync(path)) return;

      const content = this.fs.readFileSync(path, { encoding: 'utf-8' });
      cleanFiles.push({ path, content });
    });

    // 清空文件
    cleanFiles.forEach(({ path }) => {
      this.fs.writeFileSync(path, `# 由硬核树莓派 wizard 脚本清空, 内容已备份`, {
        encoding: 'utf-8',
      });
    });
  }

  /** 写入软件源 */
  private writeAptSourcesList(source: AptBaseSource) {
    mkdirp(pathUtil.dirname(this.config.aptSourceWritePath), this.fs);

    const data = [
      '# 由硬核树莓派 wizard 脚本自动生成',
      '',
      `# ${source.name} - ${source.title}`,
      ...(source.note ? [`# ${source.note}`] : []),
      ...source.toFullSourceUrls(),
    ].join('\n');

    this.fs.writeFileSync(this.config.aptSourceWritePath, data, {
      encoding: 'utf-8',
    });
  }

  update(source: AptBaseSource) {
    this.cleanAndBackUp();
    this.writeAptSourcesList(source);
  }
}
