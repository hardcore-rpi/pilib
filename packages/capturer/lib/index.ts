import { Config } from './Config';
import { Logger } from './Logger';
import { Camera, FaceDetector, Uploader } from './service';

const logger = new Logger('App');

async function run() {
  const config = new Config();
  logger.info(`config: ${config.toLogStr()}`);

  const cam = new Camera(config.CAMERA_ID, {
    height: config.CAMERA_HEIGHT,
    width: config.CAMERA_WIDTH,
  });
  const detector = new FaceDetector();
  const uploader = new Uploader(config.UPLOAD_ENDPOINT);

  const sevList = [cam, detector, uploader];
  await Promise.all(sevList.map(s => s.init()));

  logger.info('init done');

  let lastFaceCnt = 0;

  // 循环检测
  while (1) {
    const { mat } = await cam.read();
    await detector.setImg(mat);
    const { count: faceCount } = await detector.detectAllFaces();

    if (lastFaceCnt < faceCount) {
      // 有人进入画面
      logger.info(`somebody into view, ${faceCount - lastFaceCnt}`);
      // 上传
      const timestamp = new Date().valueOf();
      await uploader.upload(timestamp + '.png', mat);
    } else if (lastFaceCnt > faceCount) {
      logger.info(`somebody leave view, ${faceCount - lastFaceCnt}`);
    }

    lastFaceCnt = faceCount;
  }

  await Promise.all(sevList.map(s => s.init()));
}

run().catch(e => {
  logger.error('crashed:' + e);
  process.exit(1);
});
