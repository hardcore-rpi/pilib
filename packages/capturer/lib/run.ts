#!/usr/bin/env node

import { Config } from './Config';
import { Logger } from './Logger';
import { Camera, FaceDetector, Uploader } from './service';

const logger = new Logger('App');

async function run() {
  let exitFlag = false;

  // 监听 ctrl-c 退出
  process.on('SIGINT', () => (exitFlag = true));
  process.on('SIGABRT', () => (exitFlag = true));

  const config = new Config();
  logger.info(`config: ${config.toLogStr()}`);

  const cam = new Camera(config.CAMERA_ID, config.CAMERA_NAME, {
    height: config.CAMERA_HEIGHT,
    width: config.CAMERA_WIDTH,
  });
  const detector = new FaceDetector();
  const uploader = new Uploader(config.UPLOAD_ENDPOINT);

  const sevList = [cam, detector, uploader];
  await Promise.all(sevList.map(s => s.init()));

  logger.info('init done');

  let lastFaceCnt = 0;

  // 循环检测
  while (!exitFlag) {
    try {
      const snapshot = await cam.read();
      await detector.setImg(snapshot.mat);

      const { count: faceCount } = await detector.detectAllFaces();

      if (lastFaceCnt < faceCount) {
        // 有人进入画面，上传
        await uploader.upload(snapshot);
      }

      if (faceCount !== lastFaceCnt) {
        logger.info(`faceCount ${lastFaceCnt} ${faceCount - lastFaceCnt}`);
        lastFaceCnt = faceCount;
      }

      // clean
      snapshot.release();
    } catch (e) {
      logger.error(e.message || e);

      // 重置，并继续执行
      lastFaceCnt = 0;
    }
  }

  await Promise.all(sevList.map(s => s.release()));
}

run()
  .then(() => {
    logger.info('exit successfully');
    process.exit(0);
  })
  .catch(e => {
    logger.error('crashed:' + e);
    process.exit(1);
  });
