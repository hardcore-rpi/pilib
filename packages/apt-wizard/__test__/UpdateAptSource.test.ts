import { Volume, createFsFromVolume } from 'memfs';
import { AptTunaSource } from '../src/lib/AptSource';
import { UpdateAptSource } from '../src/lib/UpdateAptSource';

class MockUpdateAptSource extends UpdateAptSource {
  backUpPath = '/_backUpPath';
}

describe('normal', () => {
  it('normal', () => {
    const vol = Volume.fromJSON({
      '/etc/apt/sources.list': 'aaa',
      '/etc/apt/sources.list.d/raspi.list': 'bbb',
    });
    const fs = createFsFromVolume(vol);

    const ins = new MockUpdateAptSource(
      {
        aptSourceWritePath: '/etc/apt/sources.list',
        cleanPaths: ['/etc/apt/sources.list.d/raspi.list'],
      },
      fs
    );

    ins.update(new AptTunaSource());

    expect(vol.toJSON()).toMatchSnapshot();
  });
});
