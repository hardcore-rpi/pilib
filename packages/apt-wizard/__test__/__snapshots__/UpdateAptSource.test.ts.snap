// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normal normal 1`] = `
Object {
  "/etc/apt/sources.list": "# 由硬核树莓派 wizard 脚本自动生成

# tuna - 清华大学开源软件镜像站
# https://mirrors.tuna.tsinghua.edu.cn/help/raspbian/
deb http://mirrors.tuna.tsinghua.edu.cn/raspbian/raspbian/ buster main non-free contrib
deb-src http://mirrors.tuna.tsinghua.edu.cn/raspbian/raspbian/ buster main non-free contrib
deb http://mirrors.tuna.tsinghua.edu.cn/raspberrypi/ buster main ui",
  "/etc/apt/sources.list.d/raspi.list": "# 由硬核树莓派 wizard 脚本清空, 内容已备份",
}
`;